### DISCUSSION TITLE
Sparta Share pointer Allocator Warning/Error

### DISCUSSION THREAD

**Discussion by rahul-neubla** [sparcians/map#445] [Q&A] [answered]
When I use non static Sparta share ptr allocator to allocate object sent on non-zero delay port, simulator throws warning/error at end of the simulation. 
```
Destructing
WARNING: Seems that not all of the blocks made it back.  
'sparta::SpartaSharedPointerAllocator<PointerT>::~SpartaSharedPointerAllocator() [with PointerT = nisa_archive::RawInstBits<32>]'
Allocated: 2
Returned: 1

```
After debugging a bit found that DataContainer on port retains the object until its reassigned by next object sent on the port (`DataContainer::setData`). The `data_` here is never reset to nullptr. Thus, it seems if port is deallocated after Sparta share ptr allocator, as last object send on the port is still being referenced by this `DataContainer`, allocator throws.

For now I am using static sparta allocator. But would be good to know if it's expected to use static allocator by design or this is something that needs to be fixed.

**Comment by ghost** [comment, marked as answer]
You are correct.  If you use an allocator, _it must outlive_ all containers/objects in your simulation that pull from it.  This is actually documented in the [Allocator class doxygen](https://sparcians.github.io/map/classsparta_1_1SpartaSharedPointerAllocator.html#details) (about the third paragraph down in the Detailed Description).

One idea that I've seen is that modelers will create a `sparta::TreeNode` in their simulation class that contains **all** of the allocators used in simulation.  This TreeNode is created and attached to root (`getRoot()`) during BuildTree phase and is "looked for" by `sparta::Unit` types during construction.  This allocator container class is torn down automatically when the Simulation class is destroyed (which means it's the last object alive).

```
class MySimAllocators : public sparta::TreeNode
{
public:
      static constexpr char name[] = "my_sim_allocators";

      MySimAllocators(...) : .... {  ... create allocators here ... } 

     static MySimAllocators * getMySimAllocators(sparta::TreeNode *node)
     {
         MySimAllocators * allocators = nullptr;
         if(node)
         {
             if(node->hasChild(MySimAllocators::name)) {
                 allocators = node->getChild(MySimAllocators::name)->getResourceAs<MySimAllocators>();
             }
             else {
                 return getMySimAllocators(node->getParent());
             }
         }
         return allocators;
     }

    // list of allocators
};
```

Keep in mind that this is mostly how all allocator objects work.  Mine is just more "complainy" to help you find bugs.  😉

**Reply by rahul-neubla** [reply]
Thanks, had gone through docs long time back, should have checked there first.
Allocator node at root level looks quite neat solution for the problem :)
Thanks again 👍

**Reply by ghost** [reply]
You bet!

**Comment by ghost** [comment]
Ironically, this issue just popped up in Olympia.  Here's the PR that fixes it: https://github.com/riscv-software-src/riscv-perf-model/pull/97

**Reply by rahul-neubla** [reply]
what a coincidence :)

**Reply by ghost** [reply]
🤔  Good point.  More of a coincidence than irony.   😆  

I guess I was thinking, "Don't think Olympia suffers from this."   Then I realized it does!  Funny this discussion was used to fix it.  😉
